### DISCUSSION TITLE
Simplifying the input database structure

### DISCUSSION THREAD

**Discussion by nhniina** [spine-tools/SpineOpt.jl#581] [Ideas]
The number of classes and parameters in the input database structure feels a bit overwhelming for many new users. Particularly, there seem to be very many classes related to the temporal and stochastic structure. Are there ways to simplify or clarify the structure without reducing the functionalities or making it more difficult to use the advanced features of SpineOpt?

One option could be to categorise the classes in the DB Editor. This would probably require additional features to the DB Editor itself. Some classes are there to introduce certain variables whereas some classes are "just" to store parameters. However, it is not easy to grasp that idea currently. In addition, there are multiple classes related to advanced temporal and stochastic structures as well as user constraints. Thus, there could be categories, for example, for

- parameters
- variables
- advanced temporal and stochastic structures 
- user constraints

I suppose there are cases where a class belongs to multiple categories. How to handle those cases?

**Ideas for the classes related to the temporal and stochastic structure**

1. Create a new class `group` to group connections, nodes and units with similar temporal and stochastic structures. Create also new classes `connection__group`, `node__group` and `unit__group`. Replace corresponding `connection/node/unit__...` classes with

- `group__investment_stochastic_structure`
- `group__stochastic_structure`
- `group__investment_temporal_block`
- `group__temporal_block`

Perhaps in many cases the user wants to create a specific temporal structure for all biomass nodes or all connections, nodes and units in a neighbouring country. The new class `group` could make these kind of cases easier. However, there are also cases when the user wants to create a specific structure for just one node, for example. In those cases the additional layer can decrease usability.

2. Create a new class `type` (or something similar) with fixed entities `investment`, `flow`, `on` (or something similar, corresponding to the variable names). Then, replace classes `...__investment_stochastic_structure`, 
`...__stochastic_structure`, `...__investment_temporal_block`, `...__temporal_block` with new classes `...__type__stochastic_structure` and `...__type__temporal_block`.

However, we should somehow be able to indicate for the user which are valid combinations. This can make the structure just more complicated to understand.

3. Similar to 1. but create the `group` class to group temporal blocks and stochastic structures instead of connections, nodes and units. For that purpose, create also classes `stochastic_structure__group` and `temporal_block__group`. Replace corresponding `...temporal_block/stochastic_structure` classes with

- `connection__investment_group`
- `model__default_investment_group`
- `model__default_group`
- `model__group`
- `node__investment_group`
- `node__group`
- `unit__investment_group`
- `unit__group`

4. Replace `temporal_block` and `stochastic_structure` classes totally with `stochastotemporal_structure` class, perhaps with a parameter `type` and parameter value list [`temporal data`, `stochastic data`].

5. Use Map parameters instead of temporal and stochastic classes. For example, create a `model` class parameter `temporal_blocks` with contents block01, block02, block03. Where to give the necessary data for the blocks? In the same (multi-dimensional/nested) Map or elsewhere? How to make sure that the user gives consistent data? Using upcoming features of Toolbox?

6. Could/should we change `stochastic_scenario` and `stochastic_structure__stochastic_scenario` classes somehow?

**Ideas for other classes**

- To/from nodes: Should we change the classes with `to_node` and `from_node`? Some would prefer having `direction__node`, that is, more dimensions in the classes. Some would prefer to use `input` and `output` instead of `to` and `from`.
- User constraints: Is there a way to clarify the user constraint classes? Can they somehow be combined with the `group` class (that groups connections, nodes and units or temporal blocks and stochastic structures; options 1. and 3. above)?
- Reports and outputs: In the input database, the additional layer of `report` in defining how the outputs are written can feel cumbersome. Is there a way to use entity groups for that? Maybe not?
- Remove `unit__commodity` class?

**Comment by manuelma** [comment]
Option one seems like an indirection. There is one fundamental problem with it in my opinion: as pointed out above, the simplest case in which the user only needs one temporal and stochastic structure becomes more difficult as they now need to create the intermediate 'group' layer. With two structures they would break even in terms of number of relationships they'd need to create; with three or more they would see advantages. In my experience, three or more structures aren't that common. I would prefer to favor the simpler cases because those are the ones we need to explain to users more often. If they understand the simplest, they can easily see how it scales.

**Comment by manuelma** [comment]
Here are a couple more ideas:
1. Get rid of all `{unit/node/connection}__investment_temporal_block` and keep only `{unit/node/connection}__temporal_block`;
introduce a new 'enumerated' parameter to these 3 classes, called `scope` or something with values `operation`, `investment`, `all`. The default can be `all` because most of the time the user wants to use the same structure for both investments and operations?
2. Get rid of `model__default_temporal_block`  and `model__default_investment_temporal_block` and instead introduce two new 'enumerated' parameters for `model__temporal_block` called `is_default_for_operation` and `is_default_for_investment` (or something), whith values `true` and `false` defaulting to `false`.
3. Same as two above but for `stochastic_structure`.

(With this we would be removing 8 classes.)

In sum, the strategy would be to replace a class with an enumerated parameter. Does that help? I'm not sure, but maybe the user would find it more understandable this way?

One general comment that's perhaps obvious: it doesn't look like we have a lot of redundancy at the moment, so whatever we remove from one place we would probably need to have in a different place. So simplifying the structure in this sense means finding the right place for things rather than removing stuff. For example, I think we can also get rid of `commodity` and replace it with a `node` enumerated parameter too?

**Comment by DillonJ** [comment]
@manuelma nice ideas! 

A nice thing about the ennumerated parameter to set the scope is that it allows us to add new values in the future, rather than having to add a set of new classes.

Not sure about this though:
> The default can be all because most of the time the user wants to use the same structure for both investments and operations?

Usually your investment temporal block with have a very long duration compared to the operational decisions.

What do you mean by this exactly?
> I think we can also get rid of commodity and replace it with a node enumerated parameter too?

Currently we use Commodity to specify connection physics to a group of nodes, but maybe we can do that in a different way.

**Reply by manuelma** [reply]
Thank you @DillonJ 

> Usually your investment temporal block will have a very long duration compared to the operational decisions.

You're right, I wasn't sure about that. We should be able to find a reasonable default though (I guess `operation`?)

> Currently we use Commodity to specify connection physics to a group of nodes, but maybe we can do that in a different way.

You're right again, I wasn't sure about that one either. So the commodity is almost effectively acting as a node group so we can set the physics for multiple nodes at once? Couldn't we use an actual node group for that instead?

**Comment by jkiviluo** [comment]
So, if we want to avoid 'indirection' through `group` or `type` class, we can still do away with a large number of relationships just by turning relationship_classes into parameters. Sometimes those parameters would be just a choice of an entity (e.g. node has a parameter that chooses the temporal_block to be used) and sometimes they would need to be a map (a combination of an entity and a value).

Here's what classes would still be needed:

![image](https://user-images.githubusercontent.com/40472544/234563770-ce353e07-e166-44e0-9935-ae29d5c171d4.png)

And here's a figure that also shows all parameters that would be needed (using the entity format instead of object/relationship format):

![image](https://user-images.githubusercontent.com/40472544/234565873-3853b02b-8884-4c72-b815-fdc647069ce7.png)

It requires a parameter type where the user picks from the list of entities of a specific entity_class.

The downside in comparison to a separate `group` class is that e.g. choosing the same temporal structure for a selection of nodes, units, and connections requires three different 'class groups', one for each class. Maybe this is not a big hindrance, since often it's enough to just have one temporal structure for the whole model (using the **default_temporal_block** of the `model` class).

**Comment by DillonJ** [comment]
@jkiviluo 
>It requires a parameter type where the user picks from the list of entities of a specific entity_class.

Isn't that identically equivalent to a relationship? What is the advantage? That is precisely the kind of functionality relationships provide (or now multiple dimensional entities) - connecting one entity to another... what is the advantage of altering parameter functionality to do the very same thing?

Edit: I can see that it does have the impact of decluttering the relationship tree and maybe there is benefit in that - but it does seem like we are blurring the concepts of parameters and relationships

**Reply by manuelma** [reply]
I feel the same way. If we are going to go with that, I think we need a clear definition so we can explain to our users why sometimes two entities are linked via a multi-dimensional entity, and sometimes via a parameter. It could be that multi-dimensional entities that don't have any parameters themselves (like node__temporal_block) are generally modelled using this new entity-valued parameter thing?

But even if we managed to come up with that rule/explanation, it still feels like we are trying to solve a UI problem (cluttering in the entity tree, difficulty to see the system structure) with a data structure change. As @DillonJ points out, the current data structure already has the ability to link two or more entities, why should we provide another path to do the exact same thing? I wonder if we could try and look at this issue from the UI perspective? How to redesign/extend the DB editor so the user sees whay @jkiviluo presents above, only without actually turning those classes into entity-valued parameters?

(unrelated, @DillonJ I wonder why did you start a new thread rather than replying to @jkiviluo 's thread above? Was that intentional? Because in my opinion it could help to keep the discussion more organized if we commented on ideas by replying to the corresponding thread.

**Reply by jkiviluo** [reply]
I think it's a fallacy to think that there is always a clear distinction between entities and parameters. Temporal_block is a useful entity, since it holds parameters. But it is also a property (parameter) of a model (or other entities like node). It makes much more sense to say that a node has a temporal_block property than to say that a temporal_block has a node property. Relationships don't convey that.

It is also of course about presenting the user with something more understandable. It starts with having a more reasonable number of entity classes. The much shorter list of entity classes allows the user to find the appropriate things much faster, especially for new users who don't remember everything. There may still be some classes that could be weeded out, like stochastic_scenario.

**Reply by manuelma** [reply]
But in terms of the data structure, it feels simpler to introduce a new switch to the entity class than to develop an entity-valued parameter type. The entity class already has all the functionality the entity-valued parameter would require. We'd just need an additional flag so it's presented to the user in a different way, like as a parameter in a table or something like that (and not in the tree).

**Reply by DillonJ** [reply]
@manuelma I hit the wrong reply button aparrently!

@jkiviluo I see what you are getting at... you are viewing a temporal block as something more like an alternative - a property of an entity. In this case it is a special kind of entity - and maybe what you want to do is have a field in the new entity pane we have been talking about, similarly to how alternative might appear as a new selectable field there...

However, it does feel like we are duplicating a lot of functionality. Another consideration is that the current structure allows us to relate as many temporal blocks to a unit as desired. We do this to implement varying temporal resultion depending on look ahead... so I don't see a nice way of being able to do this with parameters. 

In the new layout that @manuelma demo'd it's nice that you can quickly and easily see what temporal blocks are related to what entities. I feel like this would be impossible using parameters - you might see, for a single node, what temporal block it is connected to, but the reverse doesn't work - you can't see for a temporal block, what it's connected to.

Edit: unless we wanted to introduce a "temporal_structure" entity - that could be a collection of temporal blocks, then that could possibly work

**Comment by jkiviluo** [comment]
> @manuelma: But in terms of the data structure, it feels simpler to introduce a new switch to the entity class than to develop an entity-valued parameter type.

You're probably right. I was clouded by a thought that the new data structure will give us a handle to the dimensions inside the parameter. But that that doesn't extend to the parameter values, it's only about the class/dimension level. In any case, maybe we could have a chat to clarify all the data structure related issues.

@DillonJ: We would indeed need to retain all of that functionality. But maybe it's best done by flagging these relationship classes as special. There should then also be indication which class is to be 'parametrized', so that the UI can show `node__temporal_block` as a **temporal_block** parameter in `node` class, but not the other way around.

**Reply by manuelma** [reply]
I am thinking the new entity-alternative pane could be made so it presents this 'parameter-like' relationships nicely. For example

| class | entity | alternative |
| --- | --- | --- |
| node | node1 | Base |
| node__temporal_block | node1, block1 | Base |

My idea is that `node__temporal_block` will be completely hidden in the tree - the user will never find it there. But they will see it in this new pane. So if this proves to be nice, there won't even be a need to 'parameterize' a class - we would just hide it in the tree.

**Reply by DillonJ** [reply]
@manuelma I am thinking this would be a more general entity pane, that would/could also include this functionality. For example, entities  have descriptions, meta data etc. but currently we have no nice way of viewing this and/or manipulating it easily via copy paste etc. I would see this as being a nice way to add entities in the first place. Maybe that needs its own issue to think it through properly

**Reply by manuelma** [reply]
Exactly @DillonJ that's what I thought too - so in this new pane we can create and view entities which means some of them can be skipped in the tree without detriment.

I wouldn't be opposed to also show and edit descriptions and metadata here but I'd rather prefer not to - just entities and their alternatives. The reason is one entity can have multiple alternatives so it will have multiple rows, and it would be awkward to have the description and metadata duplicated for each of those rows. Plus, I think entity-alternative is a higher-class functionality compared to description and metadata so it's good to let the user focus on that, without any distractions. In sum, I would like to have a dedicated space only for viewing and editing entities and their alternatives.

**Reply by DillonJ** [reply]
I would definitely like to see the description and maybe a preview of metadata - I think this can play nicely with alternatives since we could have alternative descriptions and metadata also for each alternative version of an entity. We have no other nice way to see a list of objects/entities with their descriptions and be able to copy and paste in new ones

Edit: the fundamental idea of this object/entity pane is to see objects and their static data/properties like description and metadata etc. (I created an issue here: https://github.com/spine-tools/Spine-Toolbox/issues/2102).

If we think that is not the place for entitiy alternatives, then I would consider another pane entirely as this is something I would really like to have... but I feel like it would be practical to find a way to make them work together

**Reply by jkiviluo** [reply]
I would be vary of having metadata in this particular entity pane. Metadata is complicated - one metadata entry can concern multiple entities/alternative entities or parameters at once. Also, one entity can have multiple metadata entries. Also, this entity pane shouldn't be too wide. This system of multiple panes requires quite a lot of screen width and we have to keep it usable also on laptop screens.

I agree that we need improvements in the ways to display metadata. Current metadata pane is not completely bad though.

... and sorry, I also replied in the wrong place earlier.

**Reply by DillonJ** [reply]
I was more thinking of a quick preview or simply something you can click into to see the metadata - much like we display maps and timeseries at the moment, you have to click into it see the actual data

**Reply by jkiviluo** [reply]
Tooltip could show metadata and right-click menu could direct to a metadata editor where it would be better to manage the complications when actually adding/editing metadata.

**Comment by nnhjy** [comment]
> - Remove unit__commodity class?

Just a reminder: there is a parameter for cumulative unit flow in that class. I suppose it is still worth keeping the functionality.

**Reply by jkiviluo** [reply]
The parameter would be maintained, just in a different way (see discussion above).
